pkg/agent/nodeportlocal/k8s/npl_controller.go

```diff
@@ -325,47 +325,67 @@ func (c *NPLController) addPodIPToCache(key, podIP string) {
 	c.podToIP[key] = podIP
 }
 
+func (c *NPLController) deletePodIPFromCache(key string) {
+	c.podIPLock.Lock()
+	defer c.podIPLock.Unlock()
+	delete(c.podToIP, key)
+}
+
+func (c *NPLController) deleteAllPortRulesIfAny(podIP string) error {
+	data := c.portTable.GetDataForPodIP(podIP)
+	for _, d := range data {
+		err := c.portTable.DeleteRule(d.PodIP, int(d.PodPort))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // handleRemovePod removes rules from port table and
 // rules programmed in the system based on implementation type (e.g. IPTABLES).
 // This also removes pod annotation from pods that are not selected by service annotation.
-func (c *NPLController) handleRemovePod(key string, obj interface{}) error {
-	klog.Infof("Got delete event for Pod: %s", key)
+func (c *NPLController) handleRemovePod(key string) error {
+	klog.V(2).Infof("Got delete event for Pod: %s", key)
 	podIP, found := c.getPodIPFromCache(key)
 	if !found {
 		klog.Infof("IP address not found for Pod: %s", key)
 		return nil
 	}
-	data := c.portTable.GetDataForPodIP(podIP)
-	for _, d := range data {
-		err := c.portTable.DeleteRule(d.PodIP, int(d.PodPort))
-		if err != nil {
-			return err
-		}
-	}
 
-	if obj != nil {
-		pod := obj.(*corev1.Pod)
-		if _, exists := pod.Annotations[NPLAnnotationKey]; exists {
-			newPod := pod.DeepCopy()
-			removePodAnnotation(newPod)
-			return c.updatePodAnnotation(newPod)
-		}
+	if err := c.deleteAllPortRulesIfAny(podIP); err != nil {
+		return err
 	}
+
+	c.deletePodIPFromCache(key)
+
 	return nil
 }
 
 // handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
 func (c *NPLController) handleAddUpdatePod(key string, obj interface{}) error {
 	newPod := obj.(*corev1.Pod).DeepCopy()
-	klog.Infof("Got add/update event for pod: %s", key)
+	klog.V(2).Infof("Got add/update event for Pod: %s", key)
 
 	podIP := newPod.Status.PodIP
 	if podIP == "" {
-		klog.Infof("IP address not found for pod: %s/%s", newPod.Namespace, newPod.Name)
+		klog.Infof("IP address not set for Pod: %s/%s", newPod.Namespace, newPod.Name)
 		return nil
 	}
 	c.addPodIPToCache(key, podIP)
 
+	if !c.isNPLEnabledForServiceOfPod(obj) {
+		if err := c.deleteAllPortRulesIfAny(podIP); err != nil {
+			return err
+		}
+		if _, exists := newPod.Annotations[NPLAnnotationKey]; exists {
+			removePodAnnotation(newPod)
+			return c.updatePodAnnotation(newPod)
+		}
+		return nil
+	}
+	klog.V(2).Infof("Pod %s/%s is selected by a Service for which NodePortLocal is enabled", newPod.Namespace, newPod.Name)
```

Comment:
```suggestion
	klog.V(2).Infof("Pod %s is selected by a Service for which NodePortLocal is enabled", key)
```

Reply:
Thanks, suggestion applied. I also updated the other log messages from the same function.